### ISSUE TITLE
Option to check only strings

### ISSUE THREAD

**Issue by AlekSi** [Jason-Rev/vscode-spell-checker#3]
> It currently checks ALL text in a document.

It would be nice to have the option to check only strings (quoted string literals and comments).

**Comment by Jason-Rev** [comment]
I can see how that could be useful.

**Comment by AlekSi** [comment]
Don't get me wrong, the fact that this plugin proved a way to check camelCase method names, etc. is unique and often useful. The problem is it shows 200 warnings in just a part of a file. Sometimes I have time to add those words to the dictionary, and sometimes I don't.

**Comment by Jason-Rev** [comment]
Ahh, thank you for the clarification.  My goal is to make adding words to your personal/project dictionary a one click action.

Question: are most of the words "keywords" in a programming language or library? Or are they part of your naming convention?

**Comment by AlekSi** [comment]
Some examples:
* Language-related, Go in my case: `struct` keyword, `strconv` package name, `omitempty` tag value. In fact, this list can be generated automatically.
* Some general words, names and abbreviations: `Samsung`, `TODO`.
* Some project specific abbreviations.
* Some forms of general words: `wasn't` is marked since `wasn` is not a valid word.

**Comment by Jason-Rev** [comment]
If you get me a list of Go keywords and common library names, I'll add them.

A future enhancement is to have keywords per language (Go, PHP, C#, etc.) that way someone can program in French and the keys words would still be correct.

**Comment by AlekSi** [comment]
> If you get me a list of Go keywords and common library names, I'll add them.
#8

**Comment by AlekSi** [comment]
> Some forms of general words: `wasn't` is marked since `wasn` is not a valid word.

Should I create a separate issue for that?

**Comment by Jason-Rev** [comment]
I took care of the wasn't issue by making sure the word splitter kept words together across the ' if both sides had letters.

**Comment by Jason-Rev** [comment]
Back to your original request:
Being able to only check strings.

This falls under a more general problem of being able to support different languages within the same document.  I was hoping that there was a way for vscode to tell me which part of the document a second of text falls under.  Vscode uses a set of rules for colorization.  It would be nice to tap into that, but I am not sure how.

I'll keep looking.

**Comment by db4** [comment]
I'm second to this. An option to check strings/comments only (with configurable string/comment delimiters per language ID) would be very useful indeed.

**Comment by jrysig** [comment]
I'm using your spell checker for C code and option to check only comments (and strings) would be great feature. There are to many names I cannot control and adding all to workspace dictionary is tedious work.

**Comment by jrysig** [comment]
I've just stumbled on vscode TODO Parser:
https://marketplace.visualstudio.com/items?itemName=minhthai.vscode-todo-parser

They are searching for comments using regexps:
https://github.com/kantlove/vscode-todo-parser/blob/master/src/const/RegexStrings.ts
https://github.com/kantlove/vscode-todo-parser/blob/master/src/types/LanguageType.ts#L38

**Comment by Jason3S** [comment]
The challenge here is to create a solution that can work for multiple languages.

Ideally we could use the syntax highlighter system, but there isn't an exposed interface.  In the short term, I could put is a regex solution, like you pointed out.

**Comment by jrysig** [comment]
In case of c/cpp projects the regex solution would be major improvement, because we don't mix languages inside one file. Also, when I see the TODO Parser with regex-only solution, it would help in many more environments.

Btw, did you open Visual Studio Code request for syntax highlighted interface?

**Comment by jens1o** [comment]
Are variables excluded from that check? It would be useful when variables get spell checked, it would also solve lots of typos...

**Comment by Jason3S** [comment]
This issue is starting to drift a bit.

Please check out 0.14.0 release and let me know if it solves any of your issues.

Turning on `allowCompoundWords: true` in the settings will also take care of many of the issues seen in .cpp and .c files.

**Comment by Jason3S** [comment]
It is now possible to check only strings and comments, but it is a bit difficult to turn on.

You can try it out by adding file level settings like this:

```javascript
// cSpell:includeRegExp string
// cSpell:includeRegExp CStyleComment
```

See the [Updated README](https://github.com/Jason-Rev/vscode-spell-checker/tree/master/client#excluding-and-including-text-to-be-checked)

**Comment by AlekSi** [comment]
Personally, after adding Go keywords, I no longer have the original problem. :)

**Comment by Jason3S** [comment]
Good to hear.  I'll close this issue for now.

**Comment by jrysig** [comment]
```
"cSpell.includeRegExpList": [
  "CStyleComment",
  "string"
],
```
Works fine for C project. Thank you.

**Comment by Jason3S** [comment]
I am glad to hear it.

**Comment by mbforbes** [comment]
Sorry to comment on a closed issue, but I was wondering if there is a way to achieve this (only spell checking comments and strings) without disabling spell checking for plain text documents (like txt and markdown)?

I tried the above (setting `includeRegExpList`) and it works great for code. But now all my text documents aren't spell checked because they don't have comments in them!

Can regular expressions be applied per-language via `languageSettings`? Or is there an "ignore regular expressions" setting that could be used for plaintext file types?

Many thanks for any help!

**Comment by Jason3S** [comment]
> Can regular expressions be applied per-language via `languageSettings`? Or is there an "ignore regular expressions" setting that could be used for plaintext file types?

To limit the scope of an `includeRegExpList` include in inside of a `languageSettings` section.
```json
{
    "languageSettings": [
        {
            "languageId": "python",
            "includeRegExpList": [
                      "CStyleComment",
                      "string"
             ]
        }
    ]
}
```
